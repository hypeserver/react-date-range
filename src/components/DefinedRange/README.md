#### Example: Default Labels

```jsx inside Markdown
import { useState } from 'react';

const [state, setState] = useState([
  {
    startDate: new Date(),
    endDate: null,
    key: 'selection'
  }
]);

<DefinedRange
  onChange={item => setState([item.selection])}
  ranges={state}
/>;
```
#### Example: Custom range labels

```jsx inside Markdown
import { useState, useEffect } from 'react';

const CustomStaticRangeLabelContent = ({ text }) => {
  const [currentDateString, setCurrentDateString] = useState(Date());

  useEffect(() => {
    const intervalId = setInterval(() => {
      setCurrentDateString(Date());
    }, 1000);

    return () => {
      if (intervalId) {
        clearInterval(intervalId);
      }
    };
  }, [setCurrentDateString]);

  return (
    <span>
      <i>{text}</i>
      <span className="random-date-string">
        <b>{currentDateString}</b>
      </span>
    </span>
  );
}

const renderStaticRangeLabel = () => (
  <CustomStaticRangeLabelContent text={'This is a custom label content: '} />
);

const [state, setState] = useState([
  {
    startDate: new Date(),
    endDate: null,
    key: 'selection'
  }
]);

<DefinedRange
  onChange={item => setState([item.selection])}
  ranges={state}
  renderStaticRangeLabel={renderStaticRangeLabel}
  staticRanges={[
    {
      label: 'Hoy',
      hasCustomRendering: true,
      range: () => ({
        startDate: new Date(),
        endDate: new Date()
      }),
      isSelected() {
        return true;
      }
    }
  ]}
/>;
```
