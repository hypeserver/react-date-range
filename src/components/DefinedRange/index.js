import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from '../../styles';
import { defaultInputRanges, defaultStaticRanges } from '../../defaultRanges';
import { rangeShape } from '../DayCell';
import InputRangeField from '../InputRangeField';
import cx from 'classnames';

class DefinedRange extends Component {
  constructor(props) {
    super(props);
    this.state = {
      rangeOffset: 0,
      focusedInput: -1,
    };
  }

  handleRangeChange = range => {
    const { onChange, ranges, focusedRange } = this.props;
    const selectedRange = ranges[focusedRange[0]];
    if (!onChange || !selectedRange) return;
    onChange({
      [selectedRange.key || `range${focusedRange[0] + 1}`]: { ...selectedRange, ...range },
    });
  };

  getRangeOptionValue(option) {
    const { ranges = [], focusedRange = [] } = this.props;

    if (typeof option.getCurrentValue !== 'function') {
      return '';
    }

    const selectedRange = ranges[focusedRange[0]] || {};
    return option.getCurrentValue(selectedRange) || '';
  }

  getSelectedRange(ranges, staticRange) {
    const focusedRangeIndex = ranges.findIndex(range => {
      if (!range.startDate || !range.endDate || range.disabled) return false;
      return staticRange.isSelected(range);
    });
    const selectedRange = ranges[focusedRangeIndex];
    return { selectedRange, focusedRangeIndex };
  }

  render() {
    const {
      headerContent,
      footerContent,
      onPreviewChange,
      inputRanges,
      staticRanges,
      ranges,
      renderStaticRangeLabel,
      rangeColors,
      className,
      hideStaticInputRanges,
    } = this.props;

    return (
      <>
        {!hideStaticInputRanges && (
          <div className={cx(styles.definedRangesWrapper, className)}>
            {headerContent}
            <div className={styles.staticRanges}>
              {staticRanges.map((staticRange, i) => {
                const { selectedRange, focusedRangeIndex } = this.getSelectedRange(ranges, staticRange);
                let labelContent;

                if (staticRange.hasCustomRendering) {
                  labelContent = renderStaticRangeLabel(staticRange);
                } else {
                  labelContent = staticRange.label;
                }

                return (
                  <button
                    type="button"
                    className={cx(styles.staticRange, {
                      [styles.staticRangeSelected]: Boolean(selectedRange),
                    })}
                    style={{
                      color: selectedRange
                        ? selectedRange.color || rangeColors[focusedRangeIndex]
                        : null,
                    }}
                    key={i}
                    onClick={() => this.handleRangeChange(staticRange.range(this.props))}
                    onFocus={() => onPreviewChange && onPreviewChange(staticRange.range(this.props))}
                    onMouseOver={() =>
                      onPreviewChange && onPreviewChange(staticRange.range(this.props))
                    }
                    onMouseLeave={() => {
                      onPreviewChange && onPreviewChange();
                    }}>
                    <span tabIndex={-1} className={styles.staticRangeLabel}>
                      {labelContent}
                    </span>
                  </button>
                );
              })}
            </div>
            <div className={styles.inputRanges}>
              {inputRanges.map((rangeOption, i) => (
                <InputRangeField
                  key={i}
                  styles={styles}
                  label={rangeOption.label}
                  onFocus={() => this.setState({ focusedInput: i, rangeOffset: 0 })}
                  onBlur={() => this.setState({ rangeOffset: 0 })}
                  onChange={value => this.handleRangeChange(rangeOption.range(value, this.props))}
                  value={this.getRangeOptionValue(rangeOption)}
                />
              ))}
            </div>
            {footerContent}
          </div>
        )}
      </>
    );
  }
}

DefinedRange.propTypes = {
  inputRanges: PropTypes.array,
  staticRanges: PropTypes.array,
  hideStaticInputRanges: PropTypes.bool,
  ranges: PropTypes.arrayOf(rangeShape),
  focusedRange: PropTypes.arrayOf(PropTypes.number),
  onPreviewChange: PropTypes.func,
  onChange: PropTypes.func,
  footerContent: PropTypes.any,
  headerContent: PropTypes.any,
  rangeColors: PropTypes.arrayOf(PropTypes.string),
  className: PropTypes.string,
  renderStaticRangeLabel: PropTypes.func,
};

DefinedRange.defaultProps = {
  inputRanges: defaultInputRanges,
  staticRanges: defaultStaticRanges,
  hideStaticInputRanges: false,
  ranges: [],
  rangeColors: ['#3d91ff', '#3ecf8e', '#fed14c'],
  focusedRange: [0, 0],
};

export default DefinedRange;
