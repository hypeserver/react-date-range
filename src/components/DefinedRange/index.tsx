import React, { useState } from 'react';
import styles from '../../styles';
import { defaultInputRangesGen, defaultStaticRangesGen } from '../../defaultRanges';
import InputRangeField from '../InputRangeField';
import classnames from 'classnames';
import { InputRange, InputRangeWihLabel, isSureRange, hasRangeGen, NotFullyEmptyRange, OtherRangeProps, MaybeEmptyRange, RangeFocus, StaticRange, WeekStartsOn } from '../../types';
import { defaultInputRanges, defaultStaticRanges, DEFAULT_WEEK_STARTS_ON } from '../..';

type ComponentProps = {
  className?: string;
  focusedRange?: RangeFocus;
  footerContent?: JSX.Element;
  headerContent?: JSX.Element;
  inputRanges?: InputRangeWihLabel[];
  onChange?: (keyedRange: { [k: string]: MaybeEmptyRange; }) => void;
  onPreviewChange?: (r?: MaybeEmptyRange) => void;
  rangeColors?: string[];
  ranges?: NotFullyEmptyRange[];
  renderStaticRangeLabel?: (r: StaticRange) => JSX.Element;
  staticRanges?: StaticRange[];
  weekStartsOn?: WeekStartsOn;
}
export type DefinedRangeCP = ComponentProps;

type DefaultComponentProps = StaticRangesInputRanges & {
  ranges: NotFullyEmptyRange[];
  rangeColors: string[];
  focusedRange: RangeFocus;
  weekStartsOn: WeekStartsOn;
};
export type DefinedRangeDCP = DefaultComponentProps;

const defaultProps: DefaultComponentProps = {
  ranges: [],
  staticRanges: defaultStaticRanges,
  inputRanges: defaultInputRanges,
  rangeColors: ['#3d91ff', '#3ecf8e', '#fed14c'],
  focusedRange: [0, 0],
  weekStartsOn: DEFAULT_WEEK_STARTS_ON,
};

type StaticRangesInputRanges = {
  staticRanges: StaticRange[];
  inputRanges: InputRangeWihLabel[];
}

function getSelectedRange(ranges: MaybeEmptyRange[], staticRange: StaticRange) {
  const focusedRangeIndex = ranges.findIndex(range => {
    if (!isSureRange<OtherRangeProps>(range) || range.disabled) return false;
    return staticRange.isSelected(range);
  });
  const selectedRange = ranges[focusedRangeIndex];
  return { selectedRange, focusedRangeIndex };
}

const handleRangeChange = (range: MaybeEmptyRange, allProps: ComponentProps) => {
  const { onChange, ranges, focusedRange } = allProps;
  const frange = focusedRange || [0, 0];
  const selectedRange = ranges && ranges[frange[0]];
  if (!onChange || !selectedRange) return;
  onChange({
    [selectedRange.key || `range${frange[0] + 1}`]: { ...selectedRange, ...range },
  });
}

const getRangeOptionValue = (option: InputRange, allProps: ComponentProps) => {
    const { ranges = [], focusedRange = [] } = allProps;

    if (typeof option.getCurrentValue !== 'function') return '';

    const selectedRange = (focusedRange[0] && ranges[focusedRange[0]]) || {};
    return option.getCurrentValue(selectedRange) || '';
  }

const DefinedRange: React.FC<ComponentProps> = (props) => {
  const allProps: ComponentProps = { ...defaultProps, ...props };

  const weekStartsOn = allProps.weekStartsOn !== undefined && allProps.weekStartsOn || DEFAULT_WEEK_STARTS_ON;
  // commented out by g because unused
  // const [rangeOffset, setRangeOffset] = useState<number>(0);
  // const [focusedInput, setFocusedInput] = useState<number>(-1);
  const [rangesRespectingWeekStartsOn,] = useState<StaticRangesInputRanges>((allProps.weekStartsOn !== undefined && allProps.weekStartsOn !== DEFAULT_WEEK_STARTS_ON)
    ? {
      staticRanges: defaultStaticRangesGen({ weekStartsOn }),
      inputRanges: defaultInputRangesGen({ weekStartsOn }),
    }
    : {
      staticRanges: allProps.staticRanges || defaultStaticRanges,
      inputRanges: allProps.inputRanges || defaultInputRanges,
    }
  );

  const {
    className,
    footerContent,
    headerContent,
    onPreviewChange,
    rangeColors,
    ranges,
    renderStaticRangeLabel,
  } = allProps;

  const {
    inputRanges,
    staticRanges,
  } = rangesRespectingWeekStartsOn;

  return (
    <div className={classnames(styles.definedRangesWrapper, className)}>
      {headerContent}
      <div className={styles.staticRanges}>
        {staticRanges.map((staticRange, i) => {
          const { selectedRange, focusedRangeIndex } = getSelectedRange(ranges || [], staticRange);
          if (staticRange.hasCustomRendering && !renderStaticRangeLabel) {
            throw new Error('You should provie a renderStaticRangeLabel function when setting staticRange.hasCustomRendering = true');
          }
          if (!staticRange.label && !renderStaticRangeLabel) {
            throw new Error('Either provie a range with a label property or a renderStaticRangeLabel function to this component');
          }
          const labelContent = renderStaticRangeLabel && staticRange.hasCustomRendering
            ? renderStaticRangeLabel(staticRange)
            : staticRange.label;
          const actualStaticRange = hasRangeGen(staticRange) ? staticRange.range(allProps) : staticRange.range as MaybeEmptyRange;

          return (
            <button
              type="button"
              className={classnames(styles.staticRange, {
                [styles.staticRangeSelected]: Boolean(selectedRange),
              })}
              style={{
                color: selectedRange
                  ? selectedRange.color || (rangeColors && rangeColors[focusedRangeIndex])
                  : undefined,
              }}
              key={i}
              onClick={() => handleRangeChange(actualStaticRange, allProps)}
              onFocus={() => onPreviewChange && onPreviewChange(actualStaticRange)}
              onMouseOver={() =>
                onPreviewChange && onPreviewChange(actualStaticRange)
              }
              onMouseLeave={() => {
                onPreviewChange && onPreviewChange();
              }}>
              <span tabIndex={-1} className={styles.staticRangeLabel}>
                {labelContent}
              </span>
            </button>
          );
        })}
      </div>
      <div className={styles.inputRanges}>
        {inputRanges.map((rangeOption, i) => (
          <InputRangeField
            key={i}
            styles={styles}
            label={rangeOption.label}
            // commented out by g because unused
            // onFocus={() => { setFocusedInput(i), setRangeOffset(0) }}
            // onBlur={() => setRangeOffset(0)}
            onFocus={() => {}}
            onBlur={() => {}}
            onChange={value => handleRangeChange(rangeOption.range(value, allProps), allProps)}
            value={getRangeOptionValue(rangeOption, allProps)}
          />
        ))}
      </div>
      {footerContent}
    </div>
  );
}

export default DefinedRange;
