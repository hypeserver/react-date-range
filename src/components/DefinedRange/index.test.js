import React from 'react';
import { render } from '@testing-library/react';

import DefinedRange from '../DefinedRange';
import * as dateFns from 'date-fns';

describe('DefinedRange tests', () => {
  test('Should call "renderStaticRangeLabel" callback correct amount of times according to the "hasCustomRendering" option', () => {
    const renderStaticRangeLabel = jest.fn();

    render(
      <DefinedRange
        staticRanges={[
          {
            label: 'Dynamic Label',
            range: {},
            isSelected(range) {
              const definedRange = this.range();
              return (
                dateFns.isSameDay(range.startDate, definedRange.startDate) &&
                dateFns.isSameDay(range.endDate, definedRange.endDate)
              );
            },
            hasCustomRendering: true,
          },
          {
            label: 'Static Label',
            range: {},
            isSelected(range) {
              const definedRange = this.range();
              return (
                dateFns.isSameDay(range.startDate, definedRange.startDate) &&
                dateFns.isSameDay(range.endDate, definedRange.endDate)
              );
            },
          },
          {
            label: 'Hede',
            range: {},
            isSelected(range) {
              const definedRange = this.range();
              return (
                dateFns.isSameDay(range.startDate, definedRange.startDate) &&
                dateFns.isSameDay(range.endDate, definedRange.endDate)
              );
            },
            hasCustomRendering: true,
          },
        ]}
        renderStaticRangeLabel={renderStaticRangeLabel}
      />
    );

    expect(renderStaticRangeLabel).toHaveBeenCalledTimes(2);
  });

  test('Should render dynamic static label contents correctly', () => {
    const renderItalicLabelContent = () => (
      <i className={'italic-label-content'}>{'Italic Content'}</i>
    );
    const renderBoldLabelContent = () => <b className={'bold-label-content'}>{'Bold Content'}</b>;
    const renderSomethingElse = () => <img className={'random-image'} />;

    const renderStaticRangeLabel = function(staticRange) {
      let result;

      if (staticRange.id === 'italic') {
        result = renderItalicLabelContent();
      } else if (staticRange.id === 'bold') {
        result = renderBoldLabelContent();
      } else {
        result = renderSomethingElse();
      }

      return result;
    };

    const wrapper = render(
      <DefinedRange
        staticRanges={[
          {
            id: 'italic',
            range: {},
            isSelected(range) {
              const definedRange = this.range();
              return (
                dateFns.isSameDay(range.startDate, definedRange.startDate) &&
                dateFns.isSameDay(range.endDate, definedRange.endDate)
              );
            },
            hasCustomRendering: true,
          },
          {
            label: 'Static Label',
            range: {},
            isSelected(range) {
              const definedRange = this.range();
              return (
                dateFns.isSameDay(range.startDate, definedRange.startDate) &&
                dateFns.isSameDay(range.endDate, definedRange.endDate)
              );
            },
          },
          {
            id: 'whatever',
            range: {},
            isSelected(range) {
              const definedRange = this.range();
              return (
                dateFns.isSameDay(range.startDate, definedRange.startDate) &&
                dateFns.isSameDay(range.endDate, definedRange.endDate)
              );
            },
            hasCustomRendering: true,
          },
          {
            id: 'bold',
            range: {},
            isSelected(range) {
              const definedRange = this.range();
              return (
                dateFns.isSameDay(range.startDate, definedRange.startDate) &&
                dateFns.isSameDay(range.endDate, definedRange.endDate)
              );
            },
            hasCustomRendering: true,
          },
        ]}
        renderStaticRangeLabel={renderStaticRangeLabel}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
