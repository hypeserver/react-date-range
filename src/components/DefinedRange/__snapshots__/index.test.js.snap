// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefinedRange tests Should render dynamic static label contents correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rdrDefinedRangesWrapper"
      >
        <div
          class="rdrStaticRanges"
        >
          <button
            class="rdrStaticRange"
            type="button"
          >
            <span
              class="rdrStaticRangeLabel"
              tabindex="-1"
            >
              <i
                class="italic-label-content"
              >
                Italic Content
              </i>
            </span>
          </button>
          <button
            class="rdrStaticRange"
            type="button"
          >
            <span
              class="rdrStaticRangeLabel"
              tabindex="-1"
            >
              Static Label
            </span>
          </button>
          <button
            class="rdrStaticRange"
            type="button"
          >
            <span
              class="rdrStaticRangeLabel"
              tabindex="-1"
            >
              <img
                class="random-image"
              />
            </span>
          </button>
          <button
            class="rdrStaticRange"
            type="button"
          >
            <span
              class="rdrStaticRangeLabel"
              tabindex="-1"
            >
              <b
                class="bold-label-content"
              >
                Bold Content
              </b>
            </span>
          </button>
        </div>
        <div
          class="rdrInputRanges"
        >
          <div
            class="rdrInputRange"
          >
            <input
              class="rdrInputRangeInput"
              max="99999"
              min="0"
              placeholder="-"
              value="-"
            />
            <span>
              days up to today
            </span>
          </div>
          <div
            class="rdrInputRange"
          >
            <input
              class="rdrInputRangeInput"
              max="99999"
              min="0"
              placeholder="-"
              value="-"
            />
            <span>
              days starting today
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rdrDefinedRangesWrapper"
    >
      <div
        class="rdrStaticRanges"
      >
        <button
          class="rdrStaticRange"
          type="button"
        >
          <span
            class="rdrStaticRangeLabel"
            tabindex="-1"
          >
            <i
              class="italic-label-content"
            >
              Italic Content
            </i>
          </span>
        </button>
        <button
          class="rdrStaticRange"
          type="button"
        >
          <span
            class="rdrStaticRangeLabel"
            tabindex="-1"
          >
            Static Label
          </span>
        </button>
        <button
          class="rdrStaticRange"
          type="button"
        >
          <span
            class="rdrStaticRangeLabel"
            tabindex="-1"
          >
            <img
              class="random-image"
            />
          </span>
        </button>
        <button
          class="rdrStaticRange"
          type="button"
        >
          <span
            class="rdrStaticRangeLabel"
            tabindex="-1"
          >
            <b
              class="bold-label-content"
            >
              Bold Content
            </b>
          </span>
        </button>
      </div>
      <div
        class="rdrInputRanges"
      >
        <div
          class="rdrInputRange"
        >
          <input
            class="rdrInputRangeInput"
            max="99999"
            min="0"
            placeholder="-"
            value="-"
          />
          <span>
            days up to today
          </span>
        </div>
        <div
          class="rdrInputRange"
        >
          <input
            class="rdrInputRangeInput"
            max="99999"
            min="0"
            placeholder="-"
            value="-"
          />
          <span>
            days starting today
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
